paris-processor/src/main/java/com/airbnb/paris/processor/writers/StyleExtensionsKotlinFile.kt

```diff
@@ -62,32 +64,208 @@ internal class StyleExtensionsKotlinFile(
+                    receiver = subExtendableStyleBuilderTypeName,
                     returnType = UNIT
+            ))
+            addStatement("builder.putStyle(%T.styleable.%L[%L], %T().apply(%N).build())",
+                    rClassName,
+                    styleable.styleableResourceName,
+                    styleableChildInfo.styleableResId.code,
+                    subExtendableStyleBuilderTypeName,
+                    builderParameter
             )
-    ).build()
+        }
+    }
+
+    /*
+     * Style builder extensions to set attributes.
+     *
+     * These are available to style builders of subclasses as well.
+     *
+     * Usage is like: "view.style { padding(10) }"
+     */
+    val attrGroups = styleable.attrs.groupBy { it.styleableResId.resourceName }
+    for (groupedAttrs in attrGroups.values) {
+        rClassName!!
+
+        val nonResTargetAttrs = groupedAttrs.filter { it.targetFormat != Format.RESOURCE_ID }
+
+        val isTargetDimensionType = nonResTargetAttrs.any { it.targetFormat.isDimensionType }
+        val isTargetColorStateListType = nonResTargetAttrs.any { it.targetFormat.isColorStateListType }
+
+        val attr = if (nonResTargetAttrs.isNotEmpty()) nonResTargetAttrs.first() else groupedAttrs.first()
+        val attrResourceName = attr.styleableResId.resourceName!!
+        val baseMethodName = styleable.attrResourceNameToCamelCase(attrResourceName)
+
+        // If the target type isn't a resource: "view.style { padding(10) }"
+        if (nonResTargetAttrs.isNotEmpty()) {
+            function(baseMethodName) {
+                addKdoc(attr.javadoc.toKPoet())
+                receiver(extendableStyleBuilderTypeName)
+
+                // TODO Make sure that this works correctly when the view code is in Kotlin and already using Kotlin types
+                parameter("value", JavaTypeName.get(attr.targetType).toKPoet()) {
+                    attr.targetFormat.valueAnnotation?.let {
+                        addAnnotation(it)
+                    }
+                }
+
+                addStatement("builder.put(%T.styleable.%L[%L], value)", rClassName, styleable.styleableResourceName, attr.styleableResId.code)
+            }
+        }
+
+        // Each attribute can be set with a resource: "view.style { paddingRes(R.dimen...) }"
+        function("${baseMethodName}Res") {
+            addKdoc(attr.javadoc.toKPoet())
+            receiver(extendableStyleBuilderTypeName)
 
-    /**
+            parameter("resId", Integer.TYPE) {
+                addAnnotation(attr.targetFormat.resAnnotation)
+            }
+
+            addStatement("builder.putRes(%T.styleable.%L[%L], resId)", rClassName, styleable.styleableResourceName, attr.styleableResId.code)
+        }
+
+        // Adds a special <attribute>Dp method that automatically converts a dp value to pixels for dimensions
+        if (isTargetDimensionType) {
+            function("${baseMethodName}Dp") {
```

Comment:
@elihart thinking about removing the special `...Dp` and `@ColorInt` functions in favor of extensions as you suggested.

Unfortunately the conversion from dp to px requires a `DisplayMetrics` instance from `Resources` so it would be much less practical:
```kotlin
view.style {
    padding(10.toDp(view.resources.displayMetrics))
}
```

It could work well for a `@ColorInt -> ColorStateList` conversion:
```kotlin
textView.style {
    textColor(Color.BLUE.stateList)
}
```
Although that makes it much less discoverable and doesn't seem worth it if that's the only gain.

Reply:
ah, that dp conversion issue is a bummer.

You're right that removing the different functions makes it less discoverable. I like how discoverable it is not to start typing an attribute and see all the options pop up.

One possible way around that is to require that in app initialization Paris is initialized with a display metrics instance and it can store that information globally. `Paris.init(displayMetrics)`. Not sure that is a path you want to go down, but otherwise the overhead to convert dp is not worth it imo.

The other issue is all of the number extensions functions would pollute the whole app, which isn't terrible, but something to be cautious of.

I think it's fine to stick with the specific functions

Reply:
There is a way to access global display metrics but those could be different from the ones of the view's context. This thread goes over the issue: https://github.com/android/android-ktx/issues/132